### PR TITLE
Glossary: Dynamic Search Bar

### PR DESCRIPTION
**Pull request by aanand93**
Finished search bar. TODO: minor styling, connect it to display the searched term, fix comments Sven left on previous pull request

### CODE REVIEW

**Review by sven-gerlach — CHANGES_REQUESTED** [review]
Hi Alex,

Good stuff! Besides the few line-by-line comments, I would also do the following:
- [ ] When the search input element is focused a blue border frame appears but is missing a right border nor does not encompass the icon element. I would suggest you add a bottom, top, and right border around the icon element as well.
- [x] The width of the drop-down should encompass the icon section to the right as well
- [x] The background highlight when hovering over the search results should extend all the way out to the right and probably have a 5px margin or so to the next highlight above and beneath. Also, you may want to change the CSS cursor prop to pointer on hover.
- [x] I would remove the results box from the content flow and set the z-index to 1000 or so such that the appearance of the results box does not shift the rest of the page down
- [ ] Instead of showing no results box if there is no match I would show a default item "No results" inside the results box.
- [x] Lastly, the search results are represented using the <a> tag. Since we are not defining a href prop and they do not link to another view I would suggest you change it to divs or p elements?

Once the above is done we are good to go on this one, I think.

File: src/pages/glossary/glossaryList.js [hunk 1]
```diff
@@ -27,7 +27,9 @@ export default function GlossaryList() {
 		glossaryArray = context.glossaryTerms;
 	}
 
-	if (context.glossaryTermImg !== undefined) {
+	// without the second condition, a page refresh yields the error "glossaryArray is undefined"
+	// Because glossary terms is only set after the first render (useEffect in App)
+	if (context.glossaryTermImg !== undefined && context.glossaryTerms !== undefined) {
```
[the remark this answers is not in the document]
Comment (sven-gerlach) [reply src/pages/glossary/glossaryList.js:32]: Sorry for being unclear. I debugged this code. It caused a serious runtime error if the user clicked browser refresh when on the glossary page. I fixed it. The above is just an fyi for @bcantello.

File: src/pages/glossary/SearchBar.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+		setFilteredData([]);
+		setEnteredWord("");
+	};
+
+	/**
+	 * Upon clicking on a search result, inject values into the glossary items which are displayed at the bottom of the
+	 * page
+	 * @param value
+	 */
+	const handleSearchTermClick = (value) => {
```
[the remark this answers is not in the document]
Comment (aanand93) [reply src/pages/glossary/SearchBar.js:37]: looks great!

**Review by sven-gerlach — APPROVED** [review]
@aanand93  and @bcantello 

Hope you are well. Alex, you did an excellent job! Almost all issues were resolved. I did go over the module in a bit more detail and then implemented a whole host of new features.

**Please note the following changes:**
- remove slice method from filteredData as that arbitrarily limited the results to 24 (not sure why)
- move controlled value function for search bar outside of handleFilter function
- use onFocus and onBlur input element attributes to set isSearchBarActive state. Display the results div only if search bar is in focus and if there is some filtered data.
- Make the search results div disappear if user clicks on anything except the results div and its child nodes
- Style searchbar, remove MUI packages, replace mui icons with png files
- Beyond navigating search results with mouse pointer, users can navigate search results with arrow keys and select one with enter key which then displays the descriptive detail below
- Add a fuzzy search functionality using the Levenshtein distance function. See detailed explaantion below.
- Add 'no matches found' string as an search result output if no matches found

**Next steps**
- [ ] fix inadvertent scrolling of browser window: for some reason using the up / down keys to scroll through the search results also scrolls up / down the window. The intended effect is that it only scrolls through the results div.

**Detailed explanation of Levenshtein implementation**
If the search term is contained inside the target string, the Levenshtein distance will be exactly the delta between the length of the target string and the search term. If the Levenshtein distance is bigger than the delta between both lengths then that means that additional characters had to be either added, swapped, or deleted. Hence, if the Levenshtein distance + fuzzyChars is smaller-equal the difference in length between the target string and search term then return true.

e.g: word="long"  target="longer" => Levenshtein distance=2 => allowing for one fuzzy character implies that a Levishtein distance of 3 should still return "longer" as a search result.  Hence, if the user types in "lone" instead of "long" "longer" should still be presented as a search result.

**Comment by bcantello** [comment]
Hey guys, sorry I have been so out of the loop. Things have been crazy with the promotion at work, and then I got sick. No fun. Anyway, just getting back to some Spacelab stuff now. First of all, this is a great start! You really did a lot of work here @aanand93 and @sven-gerlach! I am sure I will have more notes once I really get into it, but a few things that jumped out on initial inspection:

1. Aesthetically I think the search bar would be better located in line with the alphabetical sorting links, but to the right hand side.
2. There are two 'x' marks that appear when you begin typing in the search field. Both seem to work to clear the field. Obviously only one is necessary. To keep everything in line and in the same place/same style, let's get rid of the smaller blue one on the left, and keep the larger grey one.
Screen Shot 2022-03-05 at 8 43 02 PM
3. Although the search functionality does work, there seems to be some funny business. For instance, in the above screen shot I typed 'the', and there are a number of results that are returned that do not include an 'h'. If I type `theyz` it returns Three Body Resonance, but not Threshold Crossing Events (TCE). Neither of them include a y or a z, so it is odd that they are returned in general, but odder that one is eliminated.
4. Is it worth thinking about having the filter start at the beginning of the word? What I mean is, if I type `t`, should it automatically filter to only terms that begin with `t`? That is how I would intuitively expect it to work, but I can see there being value in a broader search sorting since these are science terms that not everyone will know.

**Comment by bcantello** [comment]
For the most part, this is sufficiently generic to be relocated to the components folder and used elsewhere in the site. I believe the only function that is specific to the Glossary page is `handleSearchTermClick`. That could be refactored to live in a separate file though, and passed in as a parameter. Each time the search bar is used in a new context there would need to be another function added to the search bar helper file that takes care of the search term click state values for that specific usage.

**Comment by bcantello** [comment]
I ended up taking care of some the changes I commented on. Search bar has been refactored to be a reusable component, and the scss has be updated so that the search bar is not top center.

**Comment by bcantello** [comment]
Another problem:
If arrowing down through the list of potential terms provided by the search bar, the term window does not scroll with the highlighted term. For instance, If i just enter the letter `a` it will essentially make all terms available to scroll through with my mouse so that I can select one. Only 6 terms are shown in the window at any given time. If Instead of using the mouse I use the arrow keys, once I reach the end of the 6 terms, and press the down arrow to go to the seventh term, the highlight indicator showing which term I am on disappears, and the original 6 terms remain in window. If I press the down arrow 5 times beyond the 6 showing terms, I will need to press the up arrow 5 times to get back to where the highlighted portion is visible. This means the terms are present and that the arrow keys are navigating them correctly, the view window is just not tracking where the arrow key has moved the indication box to.

**Comment by DrDrewCain** [comment]
Been a bit confused about this PR, are we keeping it or are we scrapping it--if we aren't scrapping it what is the plan for it?

**Pull request merged**